[intel compiler] Fix link time error with `LLVMgold.so`

stephenmsachs opened the pull request:
fixes https://github.com/spack/spack/issues/10308
fixes https://github.com/spack/spack/issues/18606
fixes https://github.com/spack/spack/issues/17100
fixes https://github.com/spack/spack/issues/21237
fixes https://github.com/spack/spack/issues/4261

The Intel compiler will, at link time, call `ld -plugin LLVMgold.so`, which expects libraries like `libimfo.so` to be found either in the `LD_LIBRARY_PATH` or in `LLVMgold.so`s RPATH. As `LLVMgold.so` already uses RUNPATH, I used that to extend this to the necessary library locations.

Commit:
[intel compiler] Fix link time error with `LLVMgold.so`

The Intel compiler will, at link time, call `ld -plugin LLVMgold.so`, which
expects libraries like `libimfo.so` to be found either in the `LD_LIBRARY_PATH` or
in `LLVMgold.so`s RPATH.

As `LLVMgold.so` already uses RUNPATH, I used that to extend this to the
necessary library locations.

This PR should fix issues:
https://github.com/spack/spack/issues/10308
https://github.com/spack/spack/issues/18606
https://github.com/spack/spack/issues/17100
https://github.com/spack/spack/issues/21237
https://github.com/spack/spack/issues/4261

stephenmsachs commented:
Tagging maintainers @rscohn2 @danvev

rscohn2 commented on lib/spack/spack/build_systems/intel.py:
```diff
@@ -1332,5 +1333,35 @@ def uninstall_ism(self):
+            # is the location of libimf.so. If this relative location is changed
+            # in compiler releases, then we need to search for libimf.so instead
+            # of this static path.
+            rpath = "$ORIGIN/../lib:$ORIGIN/../compiler/lib/intel64_lin"
```
One concern here is that the original rpath might not be $ORIGIN/../lib in all versions. The risk of a problem is low because oneapi replaced parallel studio > 1 year ago so there will not be any parallel studio updates that might change the path.

stephenmsachs replied:
OK, I can get the original rpath from `LLVMgold.so` itself. Do you think we can assume `readelf` (e.g. binutils) is present on the OS or do we need another dependency?

rscohn2 replied:
You can use patchelf:
```
(venv) rscohn1@rscohn1-mobl1:spack$ patchelf --print-rpath $(spack location -i intel-parallel-studio)/compilers_and_libraries_\
2020.1.217/linux/lib/LLVMgold.so
$ORIGIN/../lib:$ORIGIN/../compiler/lib/intel64_lin
(venv) rscohn1@rscohn1-mobl1:spack$
```

stephenmsachs replied:
Done.

rscohn2 reviewed (dismissed):
Thanks for the fix. It looks good. I reproduced the problem with 
```
spack install stream%intel@19.1.1.217
```
and verified that the change fixes it. I also verified that icc compilers in oneapi packages do not have this problem. For the past year, all updates to icc have been delivered through the intel-oneapi-compilers spack package.

Commit:
[intel compiler] Get original LLVMgold.so rpath

Commit:
[intel compiler] Pass style checks

rscohn2 reviewed (dismissed):
Looks good. Thanks for the fix.

stephenmsachs commented:
Hi @tgamblin . As discussed last week, can you please merge this without the codecov check passing.

alalazo commented:
OT: @rscohn2 I noticed that the Github handle for the other maintainer on the Intel packages is no longer valid (@danvev) Do we need to change it or delete it?

rscohn2 commented:
> OT: @rscohn2 I noticed that the Github handle for the other maintainer on the Intel packages is no longer valid (@danvev) Do we need to change it or delete it?

I will delete it in a separate commit. 3 files need to be fixed.

alalazo commented:
> I will delete it in a separate commit. 3 files need to be fixed

Thanks. Feel free to ping me when you do.

rscohn2 commented:
@alalazo: all required checks pass. Can you merge this?

alalazo reviewed (changes requested):
I am looking again at this recipe. I think adding `patchelf` as a dependency in two packages and adding its use in the base class might not be robust enough. Do we really need patchelf to install `intel` and `intel-parallel-studio`?

stephenmsachs commented:
Well, we need to make `libimf.so` available to the `LLVMgold.so` plugin at link time. The only options I have found was to add it the the libraries RPATH (with `patchelf`) or to put it into the `LD_LIBRARY_PATH` at link time. So I see two alternatives to this apporach:

- Add the `patchelf` dependency to the in the base class under `build_systems/intel.py`. I have not seen any other build_systems with dependencies, so I did not use this route.
- Add the path to `libimf.so` to the `LD_IBRARY_PATH` when the compiler is used. This would look something like this:
```
modified   lib/spack/spack/compilers/intel.py
@@ -3,8 +3,15 @@
 #
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 
+from llnl.util.filesystem import (
+    ancestor,
+    find_libraries
+    )
+import llnl.util.tty as tty
+
 from spack.compiler import Compiler, UnsupportedCompilerFlag
 from spack.version import ver
+import spack.util.executable
 
 
 class Intel(Compiler):
@@ -118,3 +125,22 @@ def fc_pic_flag(self):
     @property
     def stdcxx_libs(self):
         return ('-cxxlib', )
+
+    def setup_custom_environment(self, pkg, env):
+        # If LLVMgold.so plugin is used by `ld` when building, then the plugin
+        # needs libimf.so on it's LD_LIBRARY_PATH when linking the dependent.
+
+        super(Intel, self).setup_custom_environment(pkg, env)
+
+        cc = spack.util.executable.which_string(self.cc)
+        if find_libraries('LLVMgold', ancestor(cc, 3),
+                          shared=True, recursive=True):
+            library_list = find_libraries('libimf', ancestor(cc, 3),
+                                          shared=True, recursive=True)
+            # Remove the ia32 entries. This is also mostly ignored in
+            # build_systems/intel.py
+            library_dirs = [ d for d in library_list.directories if "ia32" not in d ]
+            ld_library_path = ':'.join(library_dirs)
+
+            env.prepend_path('LD_LIBRARY_PATH', ld_library_path)
+            tty.debug('PrependPath: LD_LIBRARY_PATH+%s' % ld_library_path)
```

alalazo commented:
> The only options I have found was to add it the the libraries RPATH (with patchelf) or to put it into the LD_LIBRARY_PATH at link time. 

Ok, that is fine. My concern is more on the fact that the dependency is used in a base class, but is declared in a derived class. That opens to possible runtime errors where `patchelf` is not found.

stephenmsachs commented:
I see that issue. We can catch some of it with the `if self._has_compilers`, but it could be in issue going forward. Do you want me to change this patch to any of the options described above?

alalazo commented on lib/spack/spack/build_systems/intel.py:
```diff
@@ -1332,5 +1333,38 @@ def uninstall_ism(self):
+            # in compiler releases, then we need to search for libimf.so instead
+            # of this static path.
+            for lib in LLVMgold_libs:
+                p = subprocess.Popen(['patchelf', '--print-rpath', lib],
+                                     stdout=subprocess.PIPE)
```
Here you can use `Executable` which is our wrapper around `subprocess`:
```suggestion
                patchelf = Executable('patchelf')
```
You can find the definition in `spack.util.executable`

stephenmsachs replied:
I am using this one now.

alalazo commented on lib/spack/spack/build_systems/intel.py:
```diff
@@ -1332,5 +1333,38 @@ def uninstall_ism(self):
+            # is the location of libimf.so. If this relative location is changed
+            # in compiler releases, then we need to search for libimf.so instead
+            # of this static path.
+            for lib in LLVMgold_libs:
```
Per the discussion in the PR, I think it would be sufficient to add a warning (or an error, depending on what you think is  more appropriate in this context) and return early:
```python
is not self.spec.satisfies('^patchelf'):
    # Warning or error
```

stephenmsachs replied:
Added an error

alalazo reviewed (changes requested):
I left a suggestion, let me know what you think of it.

Commit:
Adding @alalazo proposed changes

- Using spack.util.Executable() instead of subprocess
- Issue an error if patchelf dependency not satisfied.

Commit:
Fix formatting

alalazo reviewed (approved):
Thanks!
